src/xform/barrier_analysis.rs

```diff
@@ -0,0 +1,1297 @@
+    }
+    None
+  }
+
+  fn visit_expr(&mut self, expr: ExprRef<'_>) -> Option<()> {
+    println!("Expr: {:?}", expr.get_key());
+    match expr.get_opcode() {
+      Opcode::BlockIntrinsic { intrinsic } => {
+        if intrinsic == subcode::BlockIntrinsic::Barrier {
+          self.has_barrier = true;
+          println!("Buffered expr: {:?}", expr.get_operand_value(0).as_ref().unwrap().get_key());
+        }
+      }
+      Opcode::FIFOPush => {
+        let parent_module = expr
+          .get_parent()
+          .as_ref::<Block>(self.sys)
+          .unwrap()
+          .get_module();
+        println!(
+          "----------Push expr_op0: {:?} , expr_op1: {:?} , Parent_module: {:?}",
+          expr.get_operand_value(0).as_ref().unwrap().get_key(),
+          expr.get_operand_value(1).as_ref().unwrap().get_key(),
+          parent_module.as_ref::<Module>(self.sys).unwrap().get_name()
+        );
+      }
+      _ => {}
+    }
+    None
+  }
+
+  fn enter<'a>(&mut self, _sys: &SysBuilder) -> Option<()> {
+    for module in self.sys.module_iter(ModuleKind::Module) {
+      self.has_barrier = false;
+      self.to_rewrite = Some(module.clone());
+      self.visit_module(module.clone());
+      if self.has_barrier {
+        let mut visitor = GraphVisitor::new(self.sys);
+        self.submodule_container_map.insert(
+          module.get_key(),
+          SubModuleContainer {
+            module_name: module.get_name().to_string(),
+            sub_module_map: HashMap::new(),
+            sub_module_order: HashMap::new(),
+            caller_expr: HashMap::new(),
+            barrier_expr: HashMap::new(),
+            block_map: HashMap::new(),
+            block_expr: HashMap::new(),
+          },
+        );
+        if let Some(module) = self.to_rewrite.take() {
+          visitor.visit_module(module);
+        }
+        visitor.graph.gather_barrier_level(self.sys);
+
+        visitor.graph.gather_submodules(self.sys);
+        self
+          .submodule_container_map
+          .get_mut(&module.get_key())
+          .unwrap()
+          .sub_module_map = visitor.graph.submodule_map.clone();
+        self
+          .submodule_container_map
+          .get_mut(&module.get_key())
+          .unwrap()
+          .sub_module_order = visitor.graph.submodule_order.clone();
+        self
+          .submodule_container_map
+          .get_mut(&module.get_key())
+          .unwrap()
+          .caller_expr = visitor.graph.caller_expr.clone();
+        self
+          .submodule_container_map
+          .get_mut(&module.get_key())
+          .unwrap()
+          .barrier_expr = visitor.graph.barrier_expr.clone();
+        self
+          .submodule_container_map
+          .get_mut(&module.get_key())
+          .unwrap()
+          .block_map = visitor.graph.block_map.clone();
+        self
+          .submodule_container_map
+          .get_mut(&module.get_key())
+          .unwrap()
+          .block_expr = visitor.graph.block_expr.clone();
```

Comment:
Think to make a struct to bundle all the attributes so that you do not need to assign all these.
````
self.x = visitor.x;
self.y = visitor.y;
````
why don't you do:
````
self.bundle = visitor.bundle;
struct Bundle {
 x, y;
}
````